AK/Assertions.h

```diff
@@ -20,3 +20,15 @@ static constexpr bool TODO = false;
 #define TODO_RISCV64() VERIFY(TODO) /* NOLINT(cert-dcl03-c,misc-static-assert) No, this can't be static_assert, it's a runtime check */
 #define TODO_PPC64() VERIFY(TODO)   /* NOLINT(cert-dcl03-c,misc-static-assert) No, this can't be static_assert, it's a runtime check */
 #define TODO_PPC() VERIFY(TODO)     /* NOLINT(cert-dcl03-c,misc-static-assert) No, this can't be static_assert, it's a runtime check */
+
+#ifdef NDEBUG
+extern "C" __attribute__((noreturn)) void ak_assertion_failed(char const*);
+#    define ASSERT(expr)                                                               \
+        (__builtin_expect(!(expr), 0)                                                  \
+                ? ak_assertion_failed(#expr " at " __FILE__ ":" __stringify(__LINE__)) \
+                : (void)0)
+#    define ASSERT_NOT_REACHED ASSERT(false) /* NOLINT(cert-dcl03-c,misc-static-assert) No, this can't be static_assert, it's a runtime check */
```

Comment:
What's the value of `ASSERT_NOT_REACHED`? I have a hard time trying to figure out where it can be useful.

Reply:
Good point, we should replace it with `__builtin_unreachable()` for it to have some actual value 😅

Reply:
For when to use it, here's one pattern we do a lot where it could be useful:

```cpp
StringView to_string(MyEnum e)
{
    switch (e) {
    case MyEnum::Foo:
		return "Foo"sv;
    case MyEnum::Bar:
		return "Bar"sv;
    }
	ASSERT_NOT_REACHED();
}

Reply:
In this case we are allowing UB in the release build, right? And then I would like to see some profile where `VERIFY_NOT_REACHED` is performing badly.

Reply:
Marking it as unreachable does not mean we allow it, it just tells the compiler that it can ignore all other cases and optimize in that knowledge
It still isn't allowed as it's ub, we just don't know what will happen when we reach that state